### PR TITLE
Typed eqsat

### PR DESCRIPTION
**Pull request by Bastacyclop**

### CODE REVIEW

File: src/main/scala/rise/core/Expr.scala [hunk 1]
```diff
@@ -48,9 +48,13 @@ final case class DepApp[K <: Kind](f: Expr, x: K#T)(
 final case class Literal(d: semantics.Data) extends Expr {
   override val t: Type = d.dataType
   override def setType(t: Type): Literal =
-    throw TypeException(
-      "tried to set the type of a Literal, whose type should never be changed"
-    )
+    if (t != this.t) {
+      throw TypeException(
+        "tried to set the type of a Literal, whose type should never be changed"
```
Comment (michel-steuwer) [inline src/main/scala/rise/core/Expr.scala:53]: Can we have a unified style here: either we throw a `TypeException` (like here) or we use an assertion (like in line 68).

File: src/main/scala/rise/core/IsClosedForm.scala [hunk 2]
```diff
@@ -6,62 +6,71 @@ import rise.core.traverse._
...
-    override implicit def monad = OptionMonad
+  case class Visitor(boundV: Set[Identifier], boundT: Set[Kind.Identifier])
+    extends PureAccumulatorTraversal[(Set[Identifier], Set[Kind.Identifier])]
+  {
+    override val accumulator = PairMonoid(SetMonoid, SetMonoid)
 
-    override def identifier[I <: Identifier] : VarType => I => Option[I] = {
-      case Reference => i => if (boundV(i)) Some(i) else None
-      case _ => return_
+    override def identifier[I <: Identifier]: VarType => I => Pair[I] = vt => i => {
```
Comment (michel-steuwer) [inline src/main/scala/rise/core/IsClosedForm.scala:14]: What is a `Pair` with a single type variable?

File: src/main/scala/rise/core/substitute.scala [hunk 3]
```diff
@@ -33,6 +33,7 @@ object substitute {
             case Lambda(x, b) =>
               // See https://www.cs.cornell.edu/courses/cs3110/2019sp/textbook/interp/subst_lambda.html
               if (x =~= `for`) return_(e)
+              // FIXME: should be !(FV(expression) contains x)
```
Comment (michel-steuwer) [inline src/main/scala/rise/core/substitute.scala:36]: Can you please create an issue for this and link it from here

File: src/main/scala/rise/core/types/Constraints.scala [hunk 4]
```diff
@@ -44,8 +44,11 @@ case class NatCollectionConstraint(a: NatCollection, b: NatCollection)
 }
 
 object Constraint {
-  val canBeSubstituted : Set[Kind.Identifier] => Kind.Identifier with Kind.Explicitness => Boolean =
-    preserve => i => !(preserve.contains(i) || i.isExplicit)
+  def canBeSubstituted(preserve: Set[Kind.Identifier],
```
Comment (michel-steuwer) [inline src/main/scala/rise/core/types/Constraints.scala:47]: Why this change in style?

**Review by michel-steuwer — APPROVED** [review]
Hard for me to review the PR detailedly with over 3000 lines of new (and fairly unfamiliar code).
Nevertheless, the code looks clean and well structured + organised.
:+1:

File: src/main/scala/rise/core/IsClosedForm.scala (hunk 2 above)
Comment (Bastacyclop) [reply src/main/scala/rise/core/IsClosedForm.scala:14]: This `Pair` type is defined by the `PureAccumulatorTraversal` and pairs the accumulator with another type.

File: src/main/scala/rise/core/substitute.scala (hunk 3 above)
Comment (Bastacyclop) [reply src/main/scala/rise/core/substitute.scala:36]: https://github.com/rise-lang/shine/issues/164

File: src/main/scala/rise/core/types/Constraints.scala (hunk 4 above)
Comment (Bastacyclop) [reply src/main/scala/rise/core/types/Constraints.scala:47]: This allows function overloading for both `i: Kind.Identifier with Kind.Explicitness` and `i: TypeIdentifier`.

File: src/main/scala/rise/core/Expr.scala (hunk 1 above)
Comment (Bastacyclop) [reply src/main/scala/rise/core/Expr.scala]: done

**Review by michel-steuwer — APPROVED** [review]
👍
